framework/src/timesteppers/CompositionDT.C

```diff
@@ -60,40 +64,100 @@ CompositionDT::CompositionDT(const InputParameters & parameters)
           "lower_bound", "Failed to find a timestepper with the name '", time_stepper_name, "'");
 }
 
+template <typename Lambda>
+void
+CompositionDT::actOnTimeSteppers(Lambda && act)
+{
+  for (auto & ts : getTimeSteppers())
+    act(*ts);
+}
+
+void
+CompositionDT::init()
+{
+  actOnTimeSteppers([](auto & ts) { ts.init(); });
+}
+
+void
+CompositionDT::preExecute()
+{
+  actOnTimeSteppers([](auto & ts) { ts.preExecute(); });
+}
+
+void
+CompositionDT::preSolve()
+{
+  actOnTimeSteppers([](auto & ts) { ts.preSolve(); });
+}
+
+void
+CompositionDT::postSolve()
+{
+  actOnTimeSteppers([](auto & ts) { ts.postSolve(); });
+}
+
+void
+CompositionDT::postExecute()
+{
+  actOnTimeSteppers([](auto & ts) { ts.postExecute(); });
+}
+
+void
+CompositionDT::preStep()
+{
+  actOnTimeSteppers([](auto & ts) { ts.preStep(); });
+}
+
+void
+CompositionDT::postStep()
+{
+  actOnTimeSteppers([](auto & ts) { ts.postStep(); });
+}
+
+bool
+CompositionDT::constrainStep(Real & dt)
+{
+  bool at_sync_point = TimeStepper::constrainStep(dt);
+  const auto time_steppers = getTimeSteppers();
+  for (auto & ts : time_steppers)
+    if (ts->constrainStep(dt))
+      return true;
+  return at_sync_point;
+}
+
 Real
 CompositionDT::computeInitialDT()
 {
-  if (_has_initial_dt)
-    return _initial_dt;
-  else
-    return computeDT();
+  return _has_initial_dt ? _initial_dt : computeDT();
 }
 
 Real
 CompositionDT::computeDT()
 {
   const auto time_steppers = getTimeSteppers();
-
   // Note : compositionDT requires other active time steppers as input so no active time steppers
   // or only compositionDT is active is not allowed
-  if (time_steppers.size() < 2 && time_steppers[0] == this)
+  if (time_steppers.size() < 1)
     mooseError("No TimeStepper(s) are currently active to compute a timestep");
 
-  std::set<Real> dts;
-  std::set<Real> bound_dt;
+  std::set<std::pair<Real, TimeStepper *>, CompareFirst> dts, bound_dt;
```

Comment:
Can you add a comment as to why we have the custom comparator, and what these are used for